### PR TITLE
C2 mirna yacout c2 s2 connect to twitter

### PR DESCRIPTION
**Pull request by Mirna-Yacout**
#181

@AmrARaouf 
@Omar-Hossam 
@karimngr 
@Fattouh92 

files: authentications_controller + authentication.rb + tests > create & remove (auth) + tests for #90 and #81 + edit settings view (devise/regis/edit) + gemfile + routes

### CODE REVIEW

**Comment by abulazm22** [comment]
Mirna, please add the if condition missing from the view and the twitter callback
MT

File: arability/app/controllers/authentications_controller.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
+#encoding: UTF-8
+class AuthenticationsController < ApplicationController
+
```
Comment (omar-mohyeldin) [inline arability/app/controllers/authentications_controller.rb]: remove spaces in empty line

File: arability/app/controllers/authentications_controller.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    #  the current user for the connection
+    # params
+    #     none
+    # Success:
+    #  checks if a record is in the Authentications table: if avaialable returns and redirect
+    #  and if not creates a new record then redirect
+    # Failure:
+    #  none
+	def twitter_callback
+	  if I18n.locale==:en
```
Comment (omar-mohyeldin) [inline arability/app/controllers/authentications_controller.rb]: leave spaces between operands and operators

File: arability/app/controllers/authentications_controller.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	  	flash[:notice] = "Connected to Twitter successfully!"
+	  else I18n.locale==:ar
```
Comment (omar-mohyeldin) [inline arability/app/controllers/authentications_controller.rb]: leave spaces between operands and operators

File: arability/app/controllers/authentications_controller.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	  	flash[:notice] = "تم التواصل مع تويتر بنجاح!"
+	  end
+	  auth = request.env["omniauth.auth"]
+      authentication = Authentication.find_by_provider_and_gid(auth["provider"], current_gamer.id) || Authentication.create_with_omniauth(auth, current_gamer)
```
Comment (omar-mohyeldin) [inline arability/app/controllers/authentications_controller.rb]: divide line to have a maximum of 80 characters, and remove indentation to be alligned with rest of code

File: arability/app/controllers/authentications_controller.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      redirect_to "/gamers/edit"
```
Comment (omar-mohyeldin) [inline arability/app/controllers/authentications_controller.rb:40]: remove indentation to be alligned with rest of code

File: arability/app/controllers/authentications_controller.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      return
```
Comment (omar-mohyeldin) [inline arability/app/controllers/authentications_controller.rb:41]: remove indentation to be alligned with rest of code

File: arability/app/controllers/authentications_controller.rb [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    # Description:
+    #  removes connection already in the table
+    # params
+    #     none
+    # Success:
+    #  finds connection record, removes it correctly and redirect
+    # Failure:
+    #  doesnot find the record
+	def remove_twitter_connection
+	  if I18n.locale==:en
```
Comment (omar-mohyeldin) [inline arability/app/controllers/authentications_controller.rb]: leave space between operands and operators

File: arability/app/controllers/authentications_controller.rb [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	  	flash[:notice] = "Connection to Twitter removed successfully!"
+	  else I18n.locale==:ar
```
Comment (omar-mohyeldin) [inline arability/app/controllers/authentications_controller.rb]: leave space between operands and operators

File: arability/app/controllers/authentications_controller.rb [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    #  Twitter callback with failure to connect to Twitter upon rejection or cancellation
+    #  by the current_gamer himself or connection failure
+    # params
+    #  none
+    # Success:
+    #  redirect_to home page
+    # Failure:
+    #  none
+	def twitter_failure
+	  if I18n.locale==:en
```
Comment (omar-mohyeldin) [inline arability/app/controllers/authentications_controller.rb]: leave space between operands and operators

File: arability/app/controllers/authentications_controller.rb [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	  	flash[:notice] = "Failed to connect to Twitter"
+	  else I18n.locale==:ar
```
Comment (omar-mohyeldin) [inline arability/app/controllers/authentications_controller.rb]: leave space between operands and operators

File: arability/app/models/authentication.rb [hunk 11]
```diff
@@ -1,3 +1,73 @@
 class Authentication < ActiveRecord::Base
+  belongs_to :gamer
   attr_accessible :gid, :provider, :token, :token_secret
+
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: remove spaces in empty line

File: arability/app/models/authentication.rb [hunk 12]
```diff
@@ -1,3 +1,73 @@
 class Authentication < ActiveRecord::Base
+  belongs_to :gamer
   attr_accessible :gid, :provider, :token, :token_secret
+  
+  class << self
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: avoid using class << self and use self.method instead

File: arability/app/models/authentication.rb [hunk 13]
```diff
@@ -1,3 +1,73 @@
...
+  # Description:
+  #  This method is to remove the authentication record for the current user
+  # Parameters:
+  #  current_gamer: the record in Gamer table for the current user
+  # Success:
+  #  removes record for the current user in Authentication table
+  # Failure:
+  #  none
+  def remove_conn(current_gamer)
+    authentication = Authentication.find_by_gamer_id_and_provider(current_gamer.id, "twitter")
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: divide line to have a maximum of 80 characters

File: arability/app/models/authentication.rb [hunk 14]
```diff
@@ -1,3 +1,73 @@
...
+  # Success:
+  #  returns the list of common followers
+  # Failure:
+  #  returns nil if no authentication is found
+  def get_common_twitter_followers(current_gamer)
+    auth = Authentication.find_by_gamer_id_and_provider(current_gamer.id, "twitter")
+    if (auth.nil?)
+      return nil
+    end
+    result = JSON.parse(open("https://api.twitter.com/1/followers/ids.json?user_id=#{auth.gid}").read)
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: divide line to have a maximum of 80 characters

File: arability/app/models/authentication.rb [hunk 15]
```diff
@@ -1,3 +1,73 @@
...
+    followers = Array.new(result["ids"])
+    common = Array.new
+    i = 0
+    while i<followers.count
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: leave space between operands and operators

**Comment by ghost** [comment]
please fix the double click on cancel the connection with twitter as it leads to an exception but other wise :+1:

File: arability/app/models/authentication.rb [hunk 16]
```diff
@@ -1,3 +1,73 @@
...
+      if Authentication.exists?(:gid => followers.at(i), :provider => "twitter")
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: avoid use of :one => two and use one: two instead

File: arability/app/models/authentication.rb [hunk 17]
```diff
@@ -1,3 +1,73 @@
...
+        common.push(Authentication.find_by_gid_and_provider(followers.at(i), "twitter").gamer_id)
```
Comment (omar-mohyeldin) [inline arability/app/models/authentication.rb]: divide line to have a maximum of 80 characters

File: arability/app/views/devise/registrations/edit.html.erb [hunk 18]
```diff
@@ -1,12 +1,27 @@
+<% if I18n.locale==:en %>
```
Comment (omar-mohyeldin) [inline arability/app/views/devise/registrations/edit.html.erb]: add space between operands and operators

File: arability/app/views/devise/registrations/edit.html.erb [hunk 19]
```diff
@@ -1,12 +1,27 @@
...
+
 <em><h3 class="text-info">Edit <%= resource_name.to_s.humanize %></h3></em>
+<% else I18n.locale==:ar %>
```
Comment (omar-mohyeldin) [inline arability/app/views/devise/registrations/edit.html.erb]: add space between operands and operators

File: arability/app/views/devise/registrations/edit.html.erb [hunk 20]
```diff
@@ -1,12 +1,27 @@
...
-<div class="container well">
+<div class="tabbable"> <!-- Only required for left/right tabs -->
+  <ul class="nav nav-tabs">
+    <li class="active"><a href="#tab1" data-toggle="tab"><%= t(:my_info)%></a></li>
+    <li><a href="#tab2" data-toggle="tab"><%= t(:social_media)%></a></li>
+  </ul>
+  <div class="tab-content" style="width:1200px;">
+    <div class="tab-pane active" id="tab1">
+<div class="well">
 <%= form_for(resource, :as => resource_name, :url => registration_path(resource_name), :html => { :method => :put, :class => "form-horizontal" }) do |f| %>
```
Comment (omar-mohyeldin) [inline arability/app/views/devise/registrations/edit.html.erb:16]: avoid use of :one => two and use one: two instead

File: arability/app/views/devise/registrations/edit.html.erb [hunk 21]
```diff
@@ -1,12 +1,27 @@
...
   <% if resource.errors.any? %>
   <div class="alert"><%= devise_error_messages! %></div>
   <% end %>
 
+  <% if I18n.locale==:en %>
```
Comment (omar-mohyeldin) [inline arability/app/views/devise/registrations/edit.html.erb:21]: add space between operands and operators

File: arability/app/views/devise/registrations/edit.html.erb [hunk 22]
```diff
@@ -1,12 +1,27 @@
...
+
   <div class="container">
+    <% if I18n.locale==:en %>
```
Comment (omar-mohyeldin) [inline arability/app/views/devise/registrations/edit.html.erb:24]: add space between operands and operators

File: arability/app/views/devise/registrations/edit.html.erb [hunk 23]
```diff
@@ -1,12 +1,27 @@
...
   <div class="control-group"><%= f.label :username, :class => "control-label" %>
   <%= f.text_field :username, :placeholder => "Username", :style => "margin-left: 10px;" %></div>
```
Comment (omar-mohyeldin) [inline arability/app/views/devise/registrations/edit.html.erb:26]: avoid use of :one => two and use one: two instead

File: arability/app/views/devise/registrations/edit.html.erb [hunk 24]
```diff
@@ -19,11 +34,17 @@
   <div class="control-group"><%= f.label :password_confirmation, :class => "control-label" %>
   <%= f.password_field :password_confirmation, :placeholder => "Confirm Password", :style => "margin-left: 10px;" %></div>
 
-  <div class="control-group"><%= f.label :age, :class => "control-label" %>
-  <%= f.text_field :age, :placeholder => "Age", :style => "margin-left: 10px;" %></div>
+  <div><%= f.label :date_of_birth, :class => "control-label" %>
```
Comment (omar-mohyeldin) [inline arability/app/views/devise/registrations/edit.html.erb:37]: avoid use of :one => two and use one: two instead

File: arability/app/views/devise/registrations/edit.html.erb [hunk 25]
```diff
@@ -19,11 +34,17 @@
...
+    <%=  f.date_select :date_of_birth, {:start_year => Time.now.year, 
+        :end_year => Time.now.year - 95}, :style => "margin-left: 10px;" %></div><br>
```
Comment (omar-mohyeldin) [inline arability/app/views/devise/registrations/edit.html.erb:39]: avoid use of :one => two and use one: two instead

File: arability/app/views/devise/registrations/edit.html.erb [hunk 26]
```diff
@@ -19,11 +34,17 @@
   <div class="control-group"><%= f.label :password_confirmation, :class => "control-label" %>
   <%= f.password_field :password_confirmation, :placeholder => "Confirm Password", :style => "margin-left: 10px;" %></div>
 
-  <div class="control-group"><%= f.label :age, :class => "control-label" %>
-  <%= f.text_field :age, :placeholder => "Age", :style => "margin-left: 10px;" %></div>
+  <div><%= f.label :date_of_birth, :class => "control-label" %>
+    <%=  f.date_select :date_of_birth, {:start_year => Time.now.year,
```
Comment (omar-mohyeldin) [inline arability/app/views/devise/registrations/edit.html.erb:38]: avoid use of :one => two and use one: two instead

**Comment by omar-mohyeldin** [comment]
@Mirna-Yacout in edit settings view, the same problems are repeated, so please modify the rest of the code and correct the specified mistakes through the code

**Comment by omar-mohyeldin** [comment]
@Mirna-Yacout please modify your code

**Comment by Mirna-Yacout** [comment]
@Omar-Hossam please review code and note that in edit view my code starts from line 130

**Comment by omar-mohyeldin** [comment]
@Mirna-Yacout code reviewed

**Comment by Mirna-Yacout** [comment]
@Fattouh92 updated UML

**Comment by Mirna-Yacout** [comment]
@karimngr please review

**Comment by ghost** [comment]
@Mirna-Yacout  reviewed :+1:

**Pull request merged**